### ISSUE TITLE
Form-encoded data with `None` as the value gets passed as string "None"

### ISSUE THREAD

**Issue by pletnes** [encode/httpx#1500]
### Checklist

- [x] There are no similar issues or pull requests to fix it yet.
- [x] The bug is reproducible against the latest release and/or `master`.

### Describe the bug

When posting form data using `requests` and the form value is `None`, nothing is passed. With `httpx`, the `None` is converted to the string `"None"`.

### To reproduce

```python
import requests
import httpx

url = 'https://httpbin.org/post'
data = dict(foo=None)

resp_requests = requests.post(url, data=data)
pprint(resp_requests.json())
# In the response:  'form': {},

resp_httpx = httpx.post(url, data=data)
pprint(resp_httpx.json())
# In the response:  'form': {'foo': 'None'},
```

### Expected behavior

1. Expected behavior to match `requests`, since "broadly" the behavior is expected to match
2. Did not expect that `"None"` would be passed, rather e.g. an empty string might make sense?

### Actual behavior

The passed form-encoded data contains the string "None"

### Debugging material

None, but see code example

### Environment

Ubuntu 20.04 docker image

### Additional context

Writing a PoC to replace `requests` with `httpx` right here.
https://github.com/SAP/cloud-pysec/blob/6e3f92fefb2e0dc4f3c99aa7ed5dc2d70e6734ba/sap/xssec/security_context.py#L511
The method `request_token_for_client` passes `None` as a default for "all scopes".

**Comment by pletnes** [comment]
Do you consider this a bug, or just "behaves differently"?

**Comment by florimondmanca** [comment]
For some reason this rings a bell. Are we sure there are no previous issues about this?

Otherwise yeah I think the "None" string is a bit surprising. I would advise validating this isn't specifically linked to HTTPBin, eg inspect what the actual request body looks like.

**Comment by florimondmanca** [comment]
Not marking as a bug yet because we'd also need to validate if "null values" make sense in the context of HTTP forms at all.

If not, consider matching Requests by ignoring None fields?

Or raising a loud error as "this is ambiguous"?

**Comment by pletnes** [comment]
From what I've seen, passing empty string instead of None works fine, and round-trips with httpbin. This is a perfectly fine workaround for me. However, for compatibility with requests, converting `None` to empty string will not yield identical results, since then the empty string is in fact passed with the `POST` request.

I've verified that the empty-string workaround works for the project I am currently working on, so it is not only a httpbin issue. I used httpbin to


Some facts I've found so far:

- Replacing `None` with `""` (empty string) makes my code work again (context: I'm trying to replace requests with httpx, drop-in, for SAP/cloud-pysec). As far as I can tell, empty string is an allowed value, and might make sense as `None`. So it's not a "httpbin only" issue; I just used httpbin to make a minimal, self-contained reproduction of the issue.
- MDN has [an article on POST](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/POST)

**Comment by pletnes** [comment]
Relevant discussion for `request`, a node.js package, which argues that `None`, `null` etc are language-specific and do not make sense.
https://github.com/request/request/issues/1832

Scrapy appears to do the same as requests:

    If a value passed in this parameter is None, the field will not be included
    in the request, even if it was present in the response <form> element.

https://docs.scrapy.org/en/latest/topics/request-response.html#scrapy.http.FormRequest.from_response

**Comment by florimondmanca** [comment]
As to why this happens:

When `data=<dict>` is passed, we send it as URL-encoded form data, built through this call:

https://github.com/encode/httpx/blob/f3c29416f12bc580b18f94538abfbc41ac06cce5/httpx/_content.py#L106-L113

So our behavior essentially boils down to the stdlib doing this:

```python
>>> from urllib.parse import urlencode
>>> urlencode({"foo": None, "bar": "baz"})
'foo=None&bar=baz'
```

I think that's somewhat undefined behavior, since the [docs](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlencode read:

> Convert a mapping object or a sequence of two-element tuples, which may contain str or bytes objects, to a percent-encoded ASCII text string.

`urlencode()` officially only accepts str or bytes, but actually it accepts anything else, in which cases it just calls `str()` on it:

```python
>>> class Spy:
...     def __str__(self):
...             print('Gotcha')
...             return 'Spy()'
... 
>>> urlencode({'foo': Spy()})
Gotcha
'foo=Spy%28%29'
```

I'd be tempted to keep things as-is for simplicity (I believe issues around special-casing values like `None` have been a PITA for Requests over time), and just add a note in our Requests Compatibility guide about this. WDYT?

**Comment by florimondmanca** [comment]
Other relevant threads:

* https://github.com/psf/requests/issues/378
* #154 (for query string, rather than URL-encoded form data)
* #175

#175 resolved #154 by treating `params={"foo": None}` as `?foo=`. Which made me realize we have this small 

https://github.com/encode/httpx/blob/c725387b2deee4035bab43c564df0bbccaf99c52/httpx/_utils.py#L59-L71

Which we use in `QueryParams` here:

https://github.com/encode/httpx/blob/86964054d60da7796c65e501ba021440b14cc36c/httpx/_models.py#L450

We might consider applying the same transformation in `encode_urlencoded_data`…?

**Comment by tomchristie** [comment]
Personally I think requests' behaviour here is unintuitive.

> We might consider applying the same transformation in encode_urlencoded_data()…? That would still be different from Requests (and we'd have to document that — we already for for query params: https://www.python-httpx.org/compatibility/#query-parameters), but at least it would be consistent with what we already do for query params.

Yes that'd be my strong preference too.
